pydantic_ai_slim/pydantic_ai/models/bedrock.py

```diff
@@ -416,16 +421,31 @@ async def _map_messages(  # noqa: C901
         Groups consecutive ToolReturnPart objects into a single user message as required by Bedrock Claude/Nova models.
         """
         profile = BedrockModelProfile.from_profile(self.profile)
+        supports_caching = profile.bedrock_supports_prompt_caching
         system_prompt: list[SystemContentBlockTypeDef] = []
         bedrock_messages: list[MessageUnionTypeDef] = []
         document_count: Iterator[int] = count(1)
         for message in messages:
             if isinstance(message, ModelRequest):
-                for part in message.parts:
+                for i, part in enumerate(message.parts):
                     if isinstance(part, SystemPromptPart):
                         system_prompt.append({'text': part.content})
                     elif isinstance(part, UserPromptPart):
-                        bedrock_messages.extend(await self._map_user_prompt(part, document_count))
+                        # Handle case where UserPromptPart starts with a CachePoint and follows the SystemPromptPart
+                        cache_point_for_system_prompt, user_prompt_part = self._extract_leading_cache_point(
```

Comment:
Instead of extracting, could we have `_map_user_prompt` return `tuple[leading_cache_point: bool, list[MessageUnionTypeDef]]`?